README.md
# <img src="https://github.com/pip-services/pip-services/raw/master/design/Logo.png" alt="Pip.Services Logo" style="max-width:30%"> <br/> Redis components for Golang

This component library is a part of the [Pip.Services](https://github.com/pip-services/pip-services) project.
It contains the Redis components: RedisLock and RedisCache.

Quick Links:

* [Downloads](https://github.com/pip-services3-go/pip-services3-redis-go/blob/master/docs/Downloads.md)
* [API Reference](https://godoc.org/github.com/pip-services3-go/pip-services3-redis-go/)
* [Configuration](https://github.com/pip-services3-go/pip-services3-redis-go/blob/master/docs/Configuration.md)
* [Building and Testing](https://github.com/pip-services3-go/pip-services3-redis-go/blob/master/docs/Development.md)
* [Contributing](https://github.com/pip-services3-go/pip-services3-redis-go/blob/master/docs/Development.md#contrib)

## Acknowledgements

The library is created and maintained by **Sergey Seroukhov** and **Levichev Dmitry**.

The documentation is written by:
- **Levichev Dmitry**

CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Redis components for Golang Changelog

## <a name="1.2.1"></a> 1.2.1 (2023-01-12)

- Update dependencies

## <a name="1.2.0"></a> 1.2.0 (2021-04-19)

### Features
* **cache** move to github.com/go-redis/redis driver
* **cache** add support of redis cluster

## <a name="1.0.0"></a> 1.0.0 (2020-03-12) 

### Features
* **build** default factory components
* **cache** Redis Cache Components
* **lock** components of working with locks in Redis



test/lock/RedisLock_test.go
package test_lock

// import { ConfigParams } from 'pip-services3-commons-node';

// import { RedisLock } from '../../src/lock/RedisLock';
// import { LockFixture } from '../fixtures/LockFixture';

// suite('RedisLock', ()=> {
//     var _lock: RedisLock;
//     var _fixture: LockFixture;

//     setup((done) => {
//         let host = process.env['REDIS_SERVICE_HOST'] || 'localhost';
//         let port = process.env['REDIS_SERVICE_PORT'] || 6379;

//         _lock = new RedisLock();

//         let config = ConfigParams.fromTuples(
//             'connection.host', host,
//             'connection.port', port
//         );
//         _lock.configure(config);

//         _fixture = new LockFixture(_lock);

//         _lock.open(null, done);
//     });

//     teardown((done) => {
//         _lock.close(null, done);
//     });

//     test('Try Acquire Lock', (done) => {
//         _fixture.testTryAcquireLock(done);
//     });

//     test('Acquire Lock', (done) => {
//         _fixture.testAcquireLock(done);
//     });

//     test('Release Lock', (done) => {
//         _fixture.testReleaseLock(done);
//     });

// });


test/fixture/CacheFixture.go
package test_fixture

import (
	"testing"
	"time"

	ccache "github.com/pip-services3-go/pip-services3-components-go/cache"
	"github.com/stretchr/testify/assert"
)

const (
	KEY1   string = "key1"
	KEY2   string = "key2"
	VALUE1 string = "value1"
	VALUE2 string = "value2"
)

type CacheFixture struct {
	cache ccache.ICache
}

func NewCacheFixture(cache ccache.ICache) *CacheFixture {
	c := CacheFixture{}
	c.cache = cache
	return &c
}

func (c *CacheFixture) TestStoreAndRetrieve(t *testing.T) {

	_, err := c.cache.Store("", KEY1, VALUE1, 5000)
	assert.Nil(t, err)

	_, err = c.cache.Store("", KEY2, VALUE2, 5000)
	assert.Nil(t, err)

	select {
	case <-time.After(500 * time.Millisecond):
	}

	val, err := c.cache.Retrieve("", KEY1)

	assert.Nil(t, err)
	assert.NotNil(t, val)
	assert.Equal(t, VALUE1, val)

	var str string
	result, err := c.cache.RetrieveAs("", KEY2, &str)
	assert.NotNil(t, result)
	assert.Nil(t, err)
	assert.Equal(t, VALUE2, str)

}

func (c *CacheFixture) TestRetrieveExpired(t *testing.T) {

	_, err := c.cache.Store("", KEY1, VALUE1, 1000)
	assert.Nil(t, err)

	select {
	case <-time.After(1500 * time.Millisecond):
	}

	val, err := c.cache.Retrieve("", KEY1)
	assert.Nil(t, err)
	assert.Nil(t, val)

	var str string
	result, err := c.cache.RetrieveAs("", KEY1, &str)
	assert.False(t, result.(bool))
	assert.Nil(t, err)

}

func (c *CacheFixture) TestRemove(t *testing.T) {

	_, err := c.cache.Store("", KEY1, VALUE1, 1000)
	assert.Nil(t, err)

	err = c.cache.Remove("", KEY1)
	assert.Nil(t, err)

	val, err := c.cache.Retrieve("", KEY1)
	assert.Nil(t, err)
	assert.Nil(t, val)

	var str string
	result, err := c.cache.RetrieveAs("", KEY1, &str)
	assert.False(t, result.(bool))
	assert.Nil(t, err)
}


test/cache/RedisCache_test.go
package test_cache

// let process = require('process');
// let assert = require('chai').assert;
// let async = require('async');

// import { Descriptor } from 'pip-services3-commons-node';
// import { ConfigParams } from 'pip-services3-commons-node';

// import { RedisCache } from '../../src/cache/RedisCache';
// import { CacheFixture } from '../fixtures/CacheFixture';

// suite('RedisCache', ()=> {
//     let _cache: RedisCache;
//     let _fixture: CacheFixture;

//     setup((done) => {
//         let host = process.env['REDIS_SERVICE_HOST'] || 'localhost';
//         let port = process.env['REDIS_SERVICE_PORT'] || 6379;

//         _cache = new RedisCache();

//         let config = ConfigParams.fromTuples(
//             'connection.host', host,
//             'connection.port', port
//         );
//         _cache.configure(config);

//         _fixture = new CacheFixture(_cache);

//         _cache.open(null, done);
//     });

//     teardown((done) => {
//         _cache.close(null, done);
//     });

//     test('Store and Retrieve', (done) => {
//         _fixture.testStoreAndRetrieve(done);
//     });

//     test('Retrieve Expired', (done) => {
//         _fixture.testRetrieveExpired(done);
//     });

//     test('Remove', (done) => {
//         _fixture.testRemove(done);
//     });

// });


lock/RedisLock.go
package lock

import (
	"strconv"
	"time"

	"github.com/gomodule/redigo/redis"
	cconf "github.com/pip-services3-go/pip-services3-commons-go/config"
	cdata "github.com/pip-services3-go/pip-services3-commons-go/data"
	cerr "github.com/pip-services3-go/pip-services3-commons-go/errors"
	cref "github.com/pip-services3-go/pip-services3-commons-go/refer"
	cauth "github.com/pip-services3-go/pip-services3-components-go/auth"
	ccon "github.com/pip-services3-go/pip-services3-components-go/connect"
	clock "github.com/pip-services3-go/pip-services3-components-go/lock"
)

/*
RedisLock are distributed lock that is implemented based on Redis in-memory database.

Configuration parameters:

  - connection(s):
    - discovery_key:         (optional) a key to retrieve the connection from IDiscovery
    - host:                  host name or IP address
    - port:                  port number
    - uri:                   resource URI or connection string with all parameters in it
  - credential(s):
    - store_key:             key to retrieve parameters from credential store
    - username:              user name (currently is not used)
    - password:              user password
  - options:
    - retrytimeout:         timeout in milliseconds to retry lock acquisition. (Default: 100)
    - retries:               number of retries (default: 3)
    - db_num:                database number in Redis  (default 0)

References:

- *:discovery:*:*:1.0        (optional) IDiscovery services to resolve connection
- *:credential-store:*:*:1.0 (optional) Credential stores to resolve credential

Example:

    lock = NewRedisRedis();
    lock.Configure(cconf.NewConfigParamsFromTuples(
      "host", "localhost",
      "port", 6379,
    ));

    err = lock.Open("123")
      ...

    result, err := lock.TryAcquireLock("123", "key1", 3000)
    if result {
    	// Processing...
    }
    err = lock.ReleaseLock("123", "key1")
    // Continue...
*/
type RedisLock struct {
	*clock.Lock
	connectionResolver *ccon.ConnectionResolver
	credentialResolver *cauth.CredentialResolver

	lockId  string
	timeout int
	//retries int
	dbNum int

	client redis.Conn
}

// NewRedisLock method are creates a new instance of this lock.
func NewRedisLock() *RedisLock {
	c := RedisLock{
		connectionResolver: ccon.NewEmptyConnectionResolver(),
		credentialResolver: cauth.NewEmptyCredentialResolver(),
		lockId:             cdata.IdGenerator.NextLong(),
		timeout:            30000,
		//retries : 3,
		dbNum:  0,
		client: nil,
	}
	c.Lock = clock.InheritLock(&c)
	return &c
}

// Configure method are configures component by passing configuration parameters.
// Parameters:
//   - config    configuration parameters to be set.
func (c *RedisLock) Configure(config *cconf.ConfigParams) {
	c.connectionResolver.Configure(config)
	c.credentialResolver.Configure(config)

	c.timeout = config.GetAsIntegerWithDefault("options.timeout", c.timeout)
	//c.retries = config.GetAsIntegerWithDefault("options.retries", c.retries)
	c.dbNum = config.GetAsIntegerWithDefault("options.db_num", c.dbNum)
	if c.dbNum > 15 || c.dbNum < 0 {
		c.dbNum = 0
	}
}

// SetReferences method are sets references to dependent components.
// Parameters:
//   - references 	references to locate the component dependencies.
func (c *RedisLock) SetReferences(references cref.IReferences) {
	c.connectionResolver.SetReferences(references)
	c.credentialResolver.SetReferences(references)
}

// IsOpen method are checks if the component is opened.
// Returns true if the component has been opened and false otherwise.
func (c *RedisLock) IsOpen() bool {
	return c.client != nil
}

// Open method are opens the component.
// Parameters:
// 	- correlationId 	(optional) transaction id to trace execution through call chain.
// Returns: error or nil no errors occured.
func (c *RedisLock) Open(correlationId string) error {
	var connection *ccon.ConnectionParams
	var credential *cauth.CredentialParams

	connection, err := c.connectionResolver.Resolve(correlationId)

	if err == nil && connection == nil {
		err = cerr.NewConfigError(correlationId, "NO_CONNECTION", "Connection is not configured")
		return err
	}

	credential, err = c.credentialResolver.Lookup(correlationId)
	if err != nil {
		return err
	}

	var url, host, port, password string
	var dialOpts []redis.DialOption = make([]redis.DialOption, 0)

	dialOpts = append(dialOpts, redis.DialConnectTimeout(time.Duration(c.timeout)*time.Millisecond))
	dialOpts = append(dialOpts, redis.DialDatabase(c.dbNum))

	if credential != nil {
		password = credential.Password()
		dialOpts = append(dialOpts, redis.DialPassword(password))
	}

	if connection.Uri() != "" {
		url = connection.Uri()
		c.client, err = redis.DialURL(url, dialOpts...)
	} else {
		host = connection.Host()
		if host == "" {
			host = "localhost"
		}
		port = strconv.FormatInt(int64(connection.Port()), 10)
		if port == "0" {
			port = "6379"
		}
		url = host + ":" + port
		c.client, err = redis.Dial("tcp", url, dialOpts...)
	}
	return err
}

// Close method are closes component and frees used resources.
// Parameters:
//  - correlationId 	(optional) transaction id to trace execution through call chain.
// Retruns: error or nil no errors occured.
func (c *RedisLock) Close(correlationId string) error {
	if c.client != nil {
		err := c.client.Close()
		c.client = nil
		if err != nil {
			return err
		}
	}
	return nil
}

func (c *RedisLock) checkOpened(correlationId string) (state bool, err error) {
	if !c.IsOpen() {
		err = cerr.NewInvalidStateError(correlationId, "NOT_OPENED", "Connection is not opened")
		return false, err
	}

	return true, nil
}

// TryAcquireLock method are makes a single attempt to acquire a lock by its key.
// It returns immediately a positive or negative result.
// Parameters:
//  - correlationId     (optional) transaction id to trace execution through call chain.
//  - key               a unique lock key to acquire.
//  - ttl               a lock timeout (time to live) in milliseconds.
// Returns: a lock result or error.
func (c *RedisLock) TryAcquireLock(correlationId string, key string, ttl int64) (result bool, err error) {
	state, err := c.checkOpened(correlationId)
	if !state {
		return false, err
	}

	res, err := redis.String(c.client.Do("SET", key, c.lockId, "NX", "PX", ttl))
	if err != nil && err == redis.ErrNil {
		return false, nil
	}
	return res == "OK", err
}

// ReleaseLock method are releases prevously acquired lock by its key.
//  - correlationId     (optional) transaction id to trace execution through call chain.
//  - key               a unique lock key to release.
// Returns: error or nil for success.
func (c *RedisLock) ReleaseLock(correlationId string, key string) error {
	state, err := c.checkOpened(correlationId)
	if !state {
		return err
	}

	// Start transaction on key
	_, err = c.client.Do("WATCH", key)
	if err != nil {
		return err
	}

	// Read and check if lock is the same
	keyId, err := redis.String(c.client.Do("GET", key))
	if err != nil && err != redis.ErrNil {
		c.client.Do("UNWATCH")
		return err
	}
	// Remove the lock if it matches
	if keyId == c.lockId {
		c.client.Send("MULTI")
		c.client.Send("DEL", key)
		_, err = c.client.Do("EXEC")
	} else { // Cancel transaction if it doesn"t match
		_, err = c.client.Do("UNWATCH")
	}
	return err
}


cache/RedisCache.go
package persistence

import (
	"encoding/json"
	"math/rand"

	"github.com/go-redis/redis"
	cconf "github.com/pip-services3-go/pip-services3-commons-go/config"
	cerr "github.com/pip-services3-go/pip-services3-commons-go/errors"
	cref "github.com/pip-services3-go/pip-services3-commons-go/refer"
	cauth "github.com/pip-services3-go/pip-services3-components-go/auth"
	ccon "github.com/pip-services3-go/pip-services3-components-go/connect"
	"strconv"
	"time"
)

/*
Distributed cache that stores values in Redis in-memory database.

Configuration parameters:

  - connection(s):
    - discovery_key:         (optional) a key to retrieve the connection from IDiscovery
    - host:                  host name or IP address
    - port:                  port number
    - uri:                   resource URI or connection string with all parameters in it
  - credential(s):
    - store_key:             key to retrieve parameters from credential store
    - username:              user name (currently is not used)
    - password:              user password
  - options:
    - retries:               number of retries (default: 3)
    - timeout:               default caching timeout in milliseconds (default: 1 minute)
    - db_num:                database number in Redis  (default 0)
    - max_size:            	 maximum number of values stored in this cache (default: 1000)
    - cluster:            	 enable redis cluster

References:

- *:discovery:*:*:1.0        (optional) IDiscovery services to resolve connection
- *:credential-store:*:*:1.0 (optional) Credential stores to resolve credential

Example:

    cache = NewRedisCache();
    cache.Configure(cconf.NewConfigParamsFromTuples(
      "host", "localhost",
      "port", 6379,
    ));

    err = cache.Open("123")
      ...

    ret, err := cache.Store("123", "key1", []byte("ABC"))
    if err != nil {
    	...
    }

    res, err := cache.Retrive("123", "key1")
    value, _ := res.([]byte)
    fmt.Println(string(value))     // Result: "ABC"
*/
type RedisCache struct {
	connectionResolver *ccon.ConnectionResolver
	credentialResolver *cauth.CredentialResolver

	timeout int
	//retries int
	dbNum     int
	isCluster bool

	client redis.UniversalClient
}

// NewRedisCache method are creates a new instance of this cache.
func NewRedisCache() *RedisCache {
	c := RedisCache{}
	c.connectionResolver = ccon.NewEmptyConnectionResolver()
	c.credentialResolver = cauth.NewEmptyCredentialResolver()
	c.timeout = 30000
	//c.retries = 3
	c.dbNum = 0
	return &c
}

// Configure method are configures component by passing configuration parameters.
//   - config    configuration parameters to be set.
func (c *RedisCache) Configure(config *cconf.ConfigParams) {
	c.connectionResolver.Configure(config)
	c.credentialResolver.Configure(config)

	c.timeout = config.GetAsIntegerWithDefault("options.timeout", c.timeout)
	//c.retries = config.GetAsIntegerWithDefault("options.retries", c.retries)
	c.dbNum = config.GetAsIntegerWithDefault("options.db_num", c.dbNum)
	if c.dbNum > 15 || c.dbNum < 0 {
		c.dbNum = 0
	}
	c.isCluster = config.GetAsBooleanWithDefault("options.cluster", c.isCluster)
}

// Sets references to dependent components.
//   - references 	references to locate the component dependencies.
func (c *RedisCache) SetReferences(references cref.IReferences) {
	c.connectionResolver.SetReferences(references)
	c.credentialResolver.SetReferences(references)
}

// Checks if the component is opened.
// Returns true if the component has been opened and false otherwise.
func (c *RedisCache) IsOpen() bool {
	return c.client != nil
}

// Open method are opens the component.
// Parameters:
//  - correlationId 	(optional) transaction id to trace execution through call chain.
// Returns: error or nil no errors occured.
func (c *RedisCache) Open(correlationId string) error {
	var (
		connection *ccon.ConnectionParams
		credential *cauth.CredentialParams
		options    redis.Options
	)

	connection, err := c.connectionResolver.Resolve(correlationId)

	if err == nil && connection == nil {
		err = cerr.NewConfigError(correlationId, "NO_CONNECTION", "Connection is not configured")
		return err
	}

	credential, err = c.credentialResolver.Lookup(correlationId)
	if err != nil {
		return err
	}
	options.DialTimeout = time.Duration(rand.Intn(c.timeout)) * time.Millisecond
	options.DB = c.dbNum

	if credential != nil {
		options.Password = credential.Password()
	}

	if connection.Uri() != "" {
		options.Addr = connection.Uri()
	} else {
		host := connection.Host()
		if host == "" {
			host = "localhost"
		}
		port := strconv.FormatInt(int64(connection.Port()), 10)
		if port == "0" {
			port = "6379"
		}
		options.Addr = host + ":" + port
	}
	if c.isCluster {
		c.client = redis.NewClusterClient(&redis.ClusterOptions{
			OnNewNode: func(client *redis.Client) {
				client = redis.NewClient(&options)
			},
			Addrs:       append([]string{}, options.Addr),
			Password:    options.Password,
			DialTimeout: options.DialTimeout,
		})
	} else {
		c.client = redis.NewClient(&options)
	}
	_, err = c.client.Ping().Result()
	return err
}

// Close method are closes component and frees used resources.
// Parameters:
//   - correlationId 	(optional) transaction id to trace execution through call chain.
// Retruns: error or nil no errors occured.
func (c *RedisCache) Close(correlationId string) error {
	if c.client != nil {
		err := c.client.Close()
		c.client = nil
		if err != nil {
			return err
		}
	}
	return nil
}

func (c *RedisCache) checkOpened(correlationId string) (state bool, err error) {
	if !c.IsOpen() {
		err = cerr.NewInvalidStateError(correlationId, "NOT_OPENED", "Connection is not opened")
		return false, err
	}

	return true, nil
}

// Retrieve method are retrieves cached value from the cache using its key.
// If value is missing in the cache or expired it returns nil.
// Parameters:
//   - correlationId     (optional) transaction id to trace execution through call chain.
//   - key               a unique value key.
//  Retruns: cached value or error.
func (c *RedisCache) Retrieve(correlationId string, key string) (value interface{}, err error) {
	state, err := c.checkOpened(correlationId)
	if !state {
		return nil, err
	}
	item, err := c.client.Get(key).Bytes()
	if err != nil {
		if err == redis.Nil {
			return nil, nil
		}
		return nil, err
	}
	if item != nil {
		var val interface{}
		err = json.Unmarshal(item, &val)
		if err != nil {
			return nil, err
		}
		return val, nil
	}
	return nil, nil
}

// Retrive cached value from the cache using its key and restore into reference object. If value is missing in the cache or expired it returns false.
// Parameters:
//   - correlationId string
//   transaction id to trace execution through call chain.
//   - key string   a unique value key.
//   - refObj       pointer to object for restore
// Returns bool, error
func (c *RedisCache) RetrieveAs(correlationId string, key string, refObj interface{}) (interface{}, error) {
	state, err := c.checkOpened(correlationId)
	if !state {
		return false, err
	}
	item, err := c.client.Get(key).Bytes()
	if err != nil {
		if err == redis.Nil {
			return false, nil
		}
		return false, err
	}
	if item != nil {
		err = json.Unmarshal(item, refObj)
		if err != nil {
			return false, err
		}
		return true, nil
	}
	return false, nil
}

// Store method are stores value in the cache with expiration time.
// Parameters:
//   - correlationId     (optional) transaction id to trace execution through call chain.
//   - key               a unique value key.
//   - value             a value to store.
//   - timeout           expiration timeout in milliseconds.
// Retruns error or nil for success
func (c *RedisCache) Store(correlationId string, key string, value interface{}, timeout int64) (result interface{}, err error) {
	state, err := c.checkOpened(correlationId)
	if !state {
		return nil, err
	}

	jsonVal, err := json.Marshal(value)
	if err != nil {
		return nil, err
	}
	tmout := time.Duration(rand.Int63n(timeout)) * time.Millisecond
	return value, c.client.Set(key, jsonVal, tmout).Err()
}

// Removes a value from the cache by its key.
// Parameters:
//   - correlationId     (optional) transaction id to trace execution through call chain.
//   - key               a unique value key.
// Returns: error or nil for success
func (c *RedisCache) Remove(correlationId string, key string) error {
	state, err := c.checkOpened(correlationId)
	if !state {
		return err
	}
	return c.client.Del(key).Err()
}
